src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+    {
+        auto op1_lens = op1->get_shape().lens();
+        auto op2_lens = op2->get_shape().lens();
+
+        std::vector<ssize_t> dims1{static_cast<ssize_t>(calc_dim(batch_axes, op1_lens)),
+                                   -1,
+                                   static_cast<ssize_t>(calc_dim(sum_axes, op1_lens))};
+        std::vector<ssize_t> dims2{static_cast<ssize_t>(calc_dim(batch_axes, op2_lens)),
+                                   -1,
+                                   static_cast<ssize_t>(calc_dim(sum_axes, op2_lens))};
+
+        op1 = info.add_instruction(make_op("reshape", {{"dims", dims1}}), op1);
+        op2 = info.add_instruction(make_op("reshape", {{"dims", dims2}}), op2);
+        op2 = info.add_instruction(make_op("transpose", {{"permutation", {0, 2, 1}}}), op2);
+        instruction_ref op = info.add_instruction(make_op("dot"), op1, op2);
+
+        std::vector<size_t> new_lens(op1_lens.size(), 1);
+        for(auto i = 0; i < new_lens.size() - sum_axes.size(); ++i)
+            new_lens[i] = std::max(op1_lens[i], op2_lens[i]);
+
+        op = info.add_instruction(make_op("reshape", {{"dims", new_lens}}), op);
+
+        // compute output row
+        std::transform(cur_pair[0].begin(),
+                       cur_pair[0].end(),
+                       cur_pair[1].begin(),
+                       cur_pair[1].begin(),
+                       [](int lhs, int rhs) { return std::max(lhs, rhs); });
+        for(int a : sum_axes)
+            cur_pair[1][a] = -1;
+
+        return op;
+    }
+
+    instruction_ref finalize_output(const onnx_parser::node_info& info,
+                                    instruction_ref op,
+                                    const int_mat& map_mat,
+                                    int_mat& cur_pair) const
+    {
+        if(any_of(map_mat.back(), [](auto i) { return i >= 0; }))
+        {
+            cur_pair[1] = map_mat.back();
+            int_vec red;
+            for(int d = 0; d < map_mat[0].size(); ++d)
+            {
+                if(cur_pair[0][d] > 0 and cur_pair[1][d] == -1)
+                    red.push_back(d);
+            }
+
+            op = apply_reduce_sum_op(info, op, red, cur_pair[1]);
+        }
+
+        return transpose_squeeze(info, cur_pair, op, map_mat.back());
+    }
+
+    instruction_ref unsqueeze_transpose(const onnx_parser::node_info& info,
+                                        int_mat& cur_pair,
+                                        instruction_ref op) const
+    {
+        int_vec unsq_axes;
+        std::vector<std::tuple<int, int>> perm;
+
+        for(auto i = 0; i < cur_pair[1].size(); ++i)
+        {
+            if(cur_pair[1][i] == -1)
+                unsq_axes.push_back(i);
+            else
+                perm.push_back({cur_pair[1][i], i});
+        }
+        op = info.add_instruction(make_op("unsqueeze", {{"axes", unsq_axes}}), op);
+
+        std::sort(perm.begin(), perm.end(), [](auto lhs, auto rhs) {
```

Comment:
Use `by(std::less<>{}, [](auto x) { return std::get<0>(x); })` instead for the compare op.

Reply:
Done.